cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/UsageHandlerTest.java

```diff
@@ -0,0 +1,267 @@
+
+    Assert.assertEquals(0, getAdapterStreamUsage(adapter).size());
+    Assert.assertEquals(0, getStreamAdapterUsage(stream).size());
+
+    getUsageHandler().executeUsageDatasetOp(
+      new TransactionExecutor.Function<UsageHandler.UsageDatasetIterable, Void>() {
+        @Override
+        public Void apply(UsageHandler.UsageDatasetIterable input) throws Exception {
+          input.getUsageDataset().register(adapter, stream);
+          return null;
+        }
+      });
+
+    Assert.assertEquals(1, getAdapterStreamUsage(adapter).size());
+    Assert.assertEquals(stream, getAdapterStreamUsage(adapter).iterator().next());
+    Assert.assertEquals(1, getStreamAdapterUsage(stream).size());
+    Assert.assertEquals(adapter, getStreamAdapterUsage(stream).iterator().next());
+
+    getUsageHandler().executeUsageDatasetOp(
+      new TransactionExecutor.Function<UsageHandler.UsageDatasetIterable, Void>() {
+        @Override
+        public Void apply(UsageHandler.UsageDatasetIterable input) throws Exception {
+          input.getUsageDataset().unregister(adapter);
+          return null;
+        }
+      });
+  }
+
+  @Test
+  public void testAdapterDatasetUsage() throws Exception {
+    final Id.Adapter adapter = Id.Adapter.from("somespace", "noadapter");
+    final Id.DatasetInstance dataset = Id.DatasetInstance.from("somespace", "nods");
+
+    Assert.assertEquals(0, getAdapterDatasetUsage(adapter).size());
+    Assert.assertEquals(0, getDatasetAdapterUsage(dataset).size());
+
+    getUsageHandler().executeUsageDatasetOp(
+      new TransactionExecutor.Function<UsageHandler.UsageDatasetIterable, Void>() {
+        @Override
+        public Void apply(UsageHandler.UsageDatasetIterable input) throws Exception {
+          input.getUsageDataset().register(adapter, dataset);
+          return null;
+        }
+      });
+
+    Assert.assertEquals(1, getAdapterDatasetUsage(adapter).size());
+    Assert.assertEquals(dataset, getAdapterDatasetUsage(adapter).iterator().next());
+    Assert.assertEquals(1, getDatasetAdapterUsage(dataset).size());
+    Assert.assertEquals(adapter, getDatasetAdapterUsage(dataset).iterator().next());
+
+    getUsageHandler().executeUsageDatasetOp(
+      new TransactionExecutor.Function<UsageHandler.UsageDatasetIterable, Void>() {
+        @Override
+        public Void apply(UsageHandler.UsageDatasetIterable input) throws Exception {
+          input.getUsageDataset().unregister(adapter);
+          return null;
+        }
+      });
+  }
+
+  // app/program/adapter -> dataset/stream
+
+  private Set<Id.DatasetInstance> getAppDatasetUsage(Id.Application app) throws Exception {
+    HttpResponse response = doGet(
+      String.format("/v3/namespaces/%s/apps/%s/datasets", app.getNamespaceId(), app.getId()));
+    Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+    return GSON.fromJson(new InputStreamReader(response.getEntity().getContent()),
```

Comment:
You don't need to close the reader? Also, when creating Reader from InputStream, always need to provide charset.